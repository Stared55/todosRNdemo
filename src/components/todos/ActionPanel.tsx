import * as React from 'react';
import { useState } from 'react';
import { StyleSheet, TextInput, View } from 'react-native';
import { useDispatch } from 'react-redux';
import {
  addTodo,
  deleteTodos,
  toggleTodos,
} from 'src/store/todos/todosActions';
import { Colors } from 'src/utils/styles/colors';
import { xScale } from 'src/utils/styles/scale';
import { ITodo } from 'src/utils/types/todoTypes';
import CustomButton from '../shared/CustomButton';

export interface ActionPanelProps {
  selected: number[];
  clearSelected: () => void;
}

const ActionPanel: React.SFC<ActionPanelProps> = ({
  selected,
  clearSelected,
}) => {
  const [newTodoTitle, setNewTodoTitle] = useState('');
  const dispatch = useDispatch();

  const handleInputValue = (text: string): void => {
    setNewTodoTitle(text);
  };

  const handleClearSelected = (): void => {
    clearSelected();
  };

  const handleToggleTodo = (state: boolean): void => {
    dispatch(toggleTodos(selected, state));
    clearSelected();
  };

  const handleActionPanel = () => {
    if (newTodoTitle.trim() === '') {
      return;
    }
    const id = Math.floor(Math.random() * (9999 - 11));
    const newTodo: ITodo = {
      userId: 1,
      id,
      title: newTodoTitle,
      completed: false,
    };
    dispatch(addTodo(newTodo));
    setNewTodoTitle('');
  };

  const handleDeleteTodos = () => {
    dispatch(deleteTodos(selected));
    clearSelected();
  };

  return (
    <View style={styles.mainWrapper}>
      {selected?.length > 0 ? (
        <View style={styles.buttonsSection}>
          <CustomButton
            label="Set Done"
            handler={() => handleToggleTodo(true)}
            containerStyles={{
              width: '22 %',
            }}
          />
          <CustomButton
            label="Set To Do"
            handler={() => handleToggleTodo(false)}
            containerStyles={{
              width: '22 %',
            }}
          />
          <CustomButton
            label="Clear"
            handler={handleClearSelected}
            containerStyles={{
              width: '20 %',
            }}
          />
          <CustomButton
            label="Delete Todos"
            handler={handleDeleteTodos}
            containerStyles={{
              width: '30%',
            }}
          />
        </View>
      ) : (
        <View style={styles.inputWrapper}>
          <TextInput
            style={styles.input}
            onChangeText={(text) => handleInputValue(text)}
            value={newTodoTitle}
            placeholder="Take out the garbage ..."
          />
          <CustomButton
            label="ADD"
            handler={handleActionPanel}
            containerStyles={{
              width: '20%',
              height: 40,
              paddingVertical: 0,
            }}
          />
        </View>
      )}
    </View>
  );
};

export default ActionPanel;

const styles = StyleSheet.create({
  mainWrapper: {
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: Colors.white,
  },
  inputWrapper: {
    height: 60,
    flexDirection: 'row',
    paddingHorizontal: xScale(12),
    width: '90%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: Colors.white,
  },
  buttonsSection: {
    height: 60,
    flexDirection: 'row',
    paddingHorizontal: xScale(12),
    width: '90%',
    justifyContent: 'space-around',
    alignItems: 'center',
    backgroundColor: Colors.white,
  },
  input: {
    height: 40,
    width: '80%',
    borderColor: 'gray',
    borderWidth: 1,
    paddingLeft: 12,
  },
});
