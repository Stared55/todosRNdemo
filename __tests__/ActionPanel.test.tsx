import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import ActionPanel from 'src/components/todos/ActionPanel';

jest.mock('react-redux', () => ({
  useDispatch: () => () => jest.fn(),
  useSelector: (selector: any) =>
    selector({
      loading: false,
      todos: [],
      error: null,
    }),
}));

describe('ActionPanel', () => {
  const selected = [1, 3, 5];
  const handler = jest.fn();
  const placeholder = 'Take out the garbage ...';
  const label = 'ADD';
  let component: ShallowWrapper;
  let props: any;

  beforeEach(() => {
    component = shallow(
      <ActionPanel clearSelected={handler} selected={selected} />,
    );
    props = component.props();
    jest.clearAllMocks();
  });

  it('should match to snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should display input with placeholder', () => {
    component = shallow(<ActionPanel clearSelected={handler} selected={[]} />);
    props = component.props();
    const input = props.children.props.children[0];
    const inputPlaceholder = input.props.placeholder;
    expect(inputPlaceholder).toEqual(placeholder);
  });

  it('should display add button', () => {
    component = shallow(<ActionPanel clearSelected={handler} selected={[]} />);
    props = component.props();
    const button = props.children.props.children[1];
    const buttonLabel = button.props.label;
    expect(buttonLabel).toEqual(label);
  });

  it('should display only 4 buttons', () => {
    const buttonsLength = props.children.props.children.length;
    expect(buttonsLength).toEqual(4);
  });
});
