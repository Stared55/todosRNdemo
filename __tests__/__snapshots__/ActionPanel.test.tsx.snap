// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionPanel should match to snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flexDirection": "row",
        "height": 60,
        "justifyContent": "space-around",
        "paddingHorizontal": 25.714285714285715,
        "width": "90%",
      }
    }
  >
    <CustomButton
      containerStyles={
        Object {
          "width": "22 %",
        }
      }
      handler={[Function]}
      label="Set Done"
    />
    <CustomButton
      containerStyles={
        Object {
          "width": "22 %",
        }
      }
      handler={[Function]}
      label="Set To Do"
    />
    <CustomButton
      containerStyles={
        Object {
          "width": "20 %",
        }
      }
      handler={[Function]}
      label="Clear"
    />
    <CustomButton
      containerStyles={
        Object {
          "width": "30%",
        }
      }
      handler={[Function]}
      label="Delete Todos"
    />
  </View>
</View>
`;
